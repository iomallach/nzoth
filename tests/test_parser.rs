mod helpers;
use std::cell::RefCell;

use helpers::SnapshotErrors;
use insta;
use nzoth::{parser::Parser, source::SourceFile};

#[test]
fn test_valid_let_declarations() {
    insta::glob!("resources/parser_cases/valid_let_declarations.nz", |path| {
        let contents = std::fs::read_to_string(path).expect("Expect path exists");
        let source = RefCell::new(SourceFile::new(
            0,
            path.file_name()
                .expect("Expect a file")
                .to_str()
                .expect("No fail")
                .to_string(),
            contents.as_str(),
        ));

        let mut parser = Parser::new(&source);
        let program = parser.parse();
        assert_eq!(2, program.nodes.len());

        let output = format!("{:#?}", program);

        insta::assert_snapshot!(output);
    })
}

#[test]
fn test_valid_identifier() {
    insta::glob!("resources/parser_cases/valid_identifiers.nz", |path| {
        let contents = std::fs::read_to_string(path).expect("Expect path exists");
        let source = RefCell::new(SourceFile::new(
            0,
            path.file_name()
                .expect("Expect a file")
                .to_str()
                .expect("No fail")
                .to_string(),
            contents.as_str(),
        ));

        let mut parser = Parser::new(&source);
        let program = parser.parse();

        let output = format!("{:#?}", program);
        insta::assert_snapshot!(output);
    })
}

#[test]
fn test_valid_assignments() {
    insta::glob!("resources/parser_cases/valid_assignments.nz", |path| {
        let contents = std::fs::read_to_string(path).expect("Expect path exists");
        let source = RefCell::new(SourceFile::new(
            0,
            path.file_name()
                .expect("Expect a file")
                .to_str()
                .expect("No fail")
                .to_string(),
            contents.as_str(),
        ));

        let mut parser = Parser::new(&source);
        let program = parser.parse();

        let output = format!("{:#?}", program);
        insta::assert_snapshot!(output);
    })
}

#[test]
fn test_invalid_assignments() {
    insta::glob!("resources/parser_cases/invalid_assignments.nz", |path| {
        let contents = std::fs::read_to_string(path).expect("Expect path exists");
        let source = RefCell::new(SourceFile::new(
            0,
            path.file_name()
                .expect("Expect a file")
                .to_str()
                .expect("No fail")
                .to_string(),
            contents.as_str(),
        ));

        let mut parser = Parser::new(&source);
        _ = parser.parse();

        insta::assert_snapshot!(SnapshotErrors(parser.errors()));
    })
}
