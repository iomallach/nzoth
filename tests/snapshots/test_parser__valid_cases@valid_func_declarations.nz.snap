---
source: tests/test_parser.rs
input_file: tests/resources/parser_cases/valid_func_declarations.nz
---
Program {
    nodes: [
        Statement(
            FuncDeclaration(
                FuncDeclaration {
                    identifier: "foo",
                    paramemetrs: [
                        FuncParameter {
                            identifier: "bar",
                            ty: Name(
                                "int",
                            ),
                            span: Span {
                                start: Position {
                                    offset: 11,
                                    line: 0,
                                    column: 11,
                                },
                                end: Position {
                                    offset: 14,
                                    line: 0,
                                    column: 14,
                                },
                                file: FileId(
                                    0,
                                ),
                            },
                        },
                        FuncParameter {
                            identifier: "baz",
                            ty: Name(
                                "float",
                            ),
                            span: Span {
                                start: Position {
                                    offset: 23,
                                    line: 0,
                                    column: 23,
                                },
                                end: Position {
                                    offset: 26,
                                    line: 0,
                                    column: 26,
                                },
                                file: FileId(
                                    0,
                                ),
                            },
                        },
                    ],
                    body: Block {
                        nodes: [],
                        last_expression: Some(
                            Expression(
                                Infix(
                                    Identifier(
                                        "bar",
                                        Span {
                                            start: Position {
                                                offset: 50,
                                                line: 1,
                                                column: 2,
                                            },
                                            end: Position {
                                                offset: 53,
                                                line: 1,
                                                column: 5,
                                            },
                                            file: FileId(
                                                0,
                                            ),
                                        },
                                    ),
                                    Add,
                                    Identifier(
                                        "baz",
                                        Span {
                                            start: Position {
                                                offset: 56,
                                                line: 1,
                                                column: 8,
                                            },
                                            end: Position {
                                                offset: 59,
                                                line: 1,
                                                column: 11,
                                            },
                                            file: FileId(
                                                0,
                                            ),
                                        },
                                    ),
                                    Span {
                                        start: Position {
                                            offset: 50,
                                            line: 1,
                                            column: 2,
                                        },
                                        end: Position {
                                            offset: 53,
                                            line: 1,
                                            column: 5,
                                        },
                                        file: FileId(
                                            0,
                                        ),
                                    },
                                ),
                            ),
                        ),
                        span: Span {
                            start: Position {
                                offset: 46,
                                line: 0,
                                column: 46,
                            },
                            end: Position {
                                offset: 61,
                                line: 2,
                                column: 1,
                            },
                            file: FileId(
                                0,
                            ),
                        },
                    },
                    return_type: Some(
                        Name(
                            "float",
                        ),
                    ),
                    span: Span {
                        start: Position {
                            offset: 0,
                            line: 0,
                            column: 0,
                        },
                        end: Position {
                            offset: 3,
                            line: 0,
                            column: 3,
                        },
                        file: FileId(
                            0,
                        ),
                    },
                },
            ),
        ),
        Statement(
            FuncDeclaration(
                FuncDeclaration {
                    identifier: "foo",
                    paramemetrs: [
                        FuncParameter {
                            identifier: "bar",
                            ty: Name(
                                "int",
                            ),
                            span: Span {
                                start: Position {
                                    offset: 74,
                                    line: 4,
                                    column: 11,
                                },
                                end: Position {
                                    offset: 77,
                                    line: 4,
                                    column: 14,
                                },
                                file: FileId(
                                    0,
                                ),
                            },
                        },
                        FuncParameter {
                            identifier: "baz",
                            ty: Name(
                                "float",
                            ),
                            span: Span {
                                start: Position {
                                    offset: 86,
                                    line: 4,
                                    column: 23,
                                },
                                end: Position {
                                    offset: 89,
                                    line: 4,
                                    column: 26,
                                },
                                file: FileId(
                                    0,
                                ),
                            },
                        },
                    ],
                    body: Block {
                        nodes: [
                            Statement(
                                Return(
                                    Return {
                                        expression: Infix(
                                            Identifier(
                                                "bar",
                                                Span {
                                                    start: Position {
                                                        offset: 122,
                                                        line: 5,
                                                        column: 11,
                                                    },
                                                    end: Position {
                                                        offset: 125,
                                                        line: 5,
                                                        column: 14,
                                                    },
                                                    file: FileId(
                                                        0,
                                                    ),
                                                },
                                            ),
                                            Add,
                                            Identifier(
                                                "baz",
                                                Span {
                                                    start: Position {
                                                        offset: 128,
                                                        line: 5,
                                                        column: 17,
                                                    },
                                                    end: Position {
                                                        offset: 131,
                                                        line: 5,
                                                        column: 20,
                                                    },
                                                    file: FileId(
                                                        0,
                                                    ),
                                                },
                                            ),
                                            Span {
                                                start: Position {
                                                    offset: 122,
                                                    line: 5,
                                                    column: 11,
                                                },
                                                end: Position {
                                                    offset: 125,
                                                    line: 5,
                                                    column: 14,
                                                },
                                                file: FileId(
                                                    0,
                                                ),
                                            },
                                        ),
                                        span: Span {
                                            start: Position {
                                                offset: 115,
                                                line: 5,
                                                column: 4,
                                            },
                                            end: Position {
                                                offset: 125,
                                                line: 5,
                                                column: 14,
                                            },
                                            file: FileId(
                                                0,
                                            ),
                                        },
                                    },
                                ),
                            ),
                        ],
                        last_expression: None,
                        span: Span {
                            start: Position {
                                offset: 109,
                                line: 4,
                                column: 46,
                            },
                            end: Position {
                                offset: 134,
                                line: 6,
                                column: 1,
                            },
                            file: FileId(
                                0,
                            ),
                        },
                    },
                    return_type: Some(
                        Name(
                            "float",
                        ),
                    ),
                    span: Span {
                        start: Position {
                            offset: 63,
                            line: 4,
                            column: 0,
                        },
                        end: Position {
                            offset: 66,
                            line: 4,
                            column: 3,
                        },
                        file: FileId(
                            0,
                        ),
                    },
                },
            ),
        ),
    ],
}
